elements/lisk-chain/src/data_access/storage.ts

```diff
@@ -12,291 +12,537 @@
+		const block = await this._db.get<BlockHeaderJSON>(
+			`${DB_KEY_BLOCKS_ID}${id}`,
+		);
 		return block;
 	}
 
 	public async getBlockHeadersByIDs(
 		arrayOfBlockIds: ReadonlyArray<string>,
-	): Promise<BlockJSON[]> {
-		const blocks = await this._storage.entities.Block.get(
-			// eslint-disable-next-line camelcase
-			{ id_in: arrayOfBlockIds },
-			{ limit: arrayOfBlockIds.length },
-		);
-
+	): Promise<BlockHeaderJSON[]> {
+		const blocks = [];
+		for (const id of arrayOfBlockIds) {
+			const block = await this._db.get<BlockHeaderJSON>(
+				`${DB_KEY_BLOCKS_ID}${id}`,
+			);
+			blocks.push(block);
+		}
 		return blocks;
 	}
 
-	public async getBlockHeaderByHeight(height: number): Promise<BlockJSON> {
-		const [block] = await this._storage.entities.Block.get({ height });
-
-		return block;
+	public async getBlockHeaderByHeight(
+		height: number,
+	): Promise<BlockHeaderJSON> {
+		const stringHeight = formatInt(height);
+		const id = await this._db.get<string>(
+			`${DB_KEY_BLOCKS_HEIGHT}${stringHeight}`,
+		);
+		return this.getBlockHeaderByID(id);
 	}
 
 	public async getBlockHeadersByHeightBetween(
 		fromHeight: number,
 		toHeight: number,
-	): Promise<BlockJSON[]> {
-		const blocks = await this._storage.entities.Block.get(
-			// eslint-disable-next-line camelcase
-			{ height_gte: fromHeight, height_lte: toHeight },
-			{ limit: null, sort: 'height:desc' },
-		);
+	): Promise<BlockHeaderJSON[]> {
+		const stream = this._db.createReadStream({
+			gte: `${DB_KEY_BLOCKS_HEIGHT}${formatInt(fromHeight)}`,
+			lte: `${DB_KEY_BLOCKS_HEIGHT}${formatInt(toHeight)}`,
+			reverse: true,
+		});
+		const blockIDs = await new Promise<string[]>((resolve, reject) => {
+			const ids: string[] = [];
+			stream
+				.on('data', ({ value }) => {
+					ids.push(value);
+				})
+				.on('error', error => {
+					reject(error);
+				})
+				.on('end', () => {
+					resolve(ids);
+				});
+		});
 
-		return blocks;
+		return this.getBlockHeadersByIDs(blockIDs);
 	}
 
 	public async getBlockHeadersWithHeights(
 		heightList: ReadonlyArray<number>,
-	): Promise<BlockJSON[]> {
-		const blocks = await this._storage.entities.Block.get(
-			{
-				// eslint-disable-next-line camelcase
-				height_in: heightList,
-			},
-			{
-				sort: 'height:asc',
-				limit: heightList.length,
-			},
-		);
-
+	): Promise<BlockHeaderJSON[]> {
+		const blocks = [];
+		for (const height of heightList) {
+			const block = await this.getBlockHeaderByHeight(height);
+			blocks.push(block);
+		}
 		return blocks;
 	}
 
-	public async getLastBlockHeader(): Promise<BlockJSON> {
-		const [lastBlockHeader] = await this._storage.entities.Block.get(
-			{},
-			{ limit: 1, sort: 'height:desc' },
-		);
+	public async getLastBlockHeader(): Promise<BlockHeaderJSON> {
+		const stream = this._db.createReadStream({
+			gte: getFirstPrefix(DB_KEY_BLOCKS_HEIGHT),
+			lte: getLastPrefix(DB_KEY_BLOCKS_HEIGHT),
+			reverse: true,
+			limit: 1,
+		});
```

Comment:
Can you explain a bit what this `this._db.createReadStream` parameters are actually doing. I noticed that this `getFirstPrefix` is just adding a hexadecimal `\x00` and `\xFF` so who that read stream works with these suffixes?

Reply:
Adding a comment line above which refers to documentation of that function would be nice.

Reply:
the createReadStream will iterate through the `key-value` pair in lexicographical order.
`limit` is the limit of iteration, and `reverse` is the order of iteration between the lexicographical order
`gte` and `lte` is the key where you want to start or finish